### ISSUE TITLE
Batch decode M≥2: fused SwiGLU capture segmentation costs a fixed ~11-22 ms/step; byte-identical 4.9× fix ready but blocked on the capture-safe invariant

### ISSUE THREAD

**Issue by justinchuby** [justinchuby/onnx-genai#1334]
## Summary

Batch decode at **M≥2** pays a **fixed per-step penalty of ~11–22 ms** on a resident model, and it is **not** in attention/KV/scheduler/sampling. Measured root cause: the fused **gate/up-SwiGLU `MatMulNBits`** node has no capture-safe path at M>1, so it falls to the tiled prefill GEMM (`last_call_capture_safe=false`). That fragments the whole-subgraph CUDA graph into **25 segments with 24 eagerly-dispatched seam nodes per step**, and the eager per-node dispatch is the entire cost.

A **byte-identical, streaming-neutral fix exists and is proven (4.9× at batch-2)**, but it is **blocked** on another team's DRY invariant. This issue records the diagnosis, the ready fix, and the exact blocker so the owning work can unblock it. **Widening their DRY guard is not an acceptable resolution.**

## Hardware / model / invocation

- **Hardware:** Intel i7-13800H (14C/20T), RTX 4060 Laptop 8 GB, driver 591.55, CUDA 13.1 runtime (WDDM).
- **Model:** `qwen05b-q4` (int4, block-32, **resident** in 8 GB). Single-threaded decode loop.
- CUDA runtime on PATH:
  ```powershell
  $nv = "$env:LOCALAPPDATA\anaconda3\Lib\site-packages\nvidia"
  $env:PATH = "$nv\cu13\bin\x86_64;$nv\cudnn\bin;$env:PATH"
  ```
- Repro (per-step CUDA phase profiler + segment/seam reporting from #1324):
  ```powershell
  $env:ONNX_GENAI_PROFILE_CUDA_DECODE_STEPS = "1"
  profile_native --model <qwen05b-q4> --native-decode-batch-sweep 1,2,4 --tokens 64 `
    --prompt "..." --ep cuda
  ```

## Measured diagnosis (qwen05b-q4, resident, RTX 4060, single-threaded)

**Capture is live and healthy at every batch size** (owner's prime suspect — capture degrading to eager at M≥2 — falsified):

| batch | cuda_graph_measured | segments | seam_nodes |
|------:|---------------------|---------:|-----------|
| 1 | captures=1 replays=62 fallbacks=0 invalidations=0 | **1** | none |
| 2 | captures=1 replays=62 fallbacks=0 invalidations=0 | **25** | `MatMulNBits[KernelCaptureUnsupported]×24` |
| 4 | captures=1 replays=62 fallbacks=0 invalidations=0 | **25** | `MatMulNBits[KernelCaptureUnsupported]×24` |

**Per-step phase breakdown (medians, ms) — the entire M=1→M=2 delta attributed:**

| phase | M=1 | M=2 | delta |
|-------|----:|----:|------:|
| total | 2.68 | 13.06 | **+10.38** |
| **kernel_host_dispatch** | **0.00** | **11.28** | **+11.28** |
| executor_other | 0.24 | 0.89 | +0.65 |
| run_unattributed | 0.24 | 0.94 | +0.70 |
| logits_read_sync | 2.43 | 0.85 | −1.58 |
| build_inputs / capture_check / finite_check / admit_sync | ~0 | ~0 | ~0 |

The whole jump is `kernel_host_dispatch` (the per-node executor eagerly re-running the 24 seam nodes every step), partially offset by `logits_read_sync` dropping 1.58 ms. Attention, KV, scheduler, and sampling contribute **~0** to the delta.

**Classification: fixed per-step, NOT per-row.** Median 20.2 ms at M=2 vs **19.5 ms at M=4** — flat in M. The batch pays this once per step for being a batch, independent of M. It is **structural (capture segmentation)**, not diffuse, and therefore fixable. (A longer-context run measured the flat term at ~22 ms; short-context here ~11 ms — same mechanism, context-dependent magnitude.)

## The fix (ready, proven, held)

- Branch: **`squad/batch-swiglu-capture-fix`** (`e91acd02`).
- Approach: capture-safe **looped M==1 gate/up-SwiGLU GEMV per row** for small M, mirroring merged #1312. Segments **25 → 1** at all N.
- **Byte-identical:** each row of the M-batched path is bit-for-bit (`to_bits`) equal to that row run alone as M==1, across shapes (`row_identity` + `cross_identity` = true).
- **Streaming-neutral on 14B:** `htod_bytes_per_token` byte-identical with the loop on and off on `qwen14b-zp` — the weight-streaming mechanism is untouched.
- **Result:** predicted ~5 ms → **achieved 5.15 ms/step at batch-2 (4.9×)**, aggregate 80 → 388 tok/s; batch-4 2.5×; batch-8 1.7×.
- **Backend-neutral / DRY-satisfied:** the fix is in the **shared CUDA EP kernel** (`onnx-runtime-ep-cuda/src/kernels/matmul_nbits.rs`, `run_f16_gate_up_swiglu`), so **both native and ORT decode inherit it** — no native-only path.

## The exact blocker

The fix advertises `last_call_capture_safe = true` for small M, which violates the invariant asserted by the swiglu team's bit-exact tests:

- `"only M=1 decode may be advertised capture-safe"` at **`crates/onnx-runtime-ep-cuda/src/kernels/matmul_nbits.rs:14561`** and **`:14962`**.
- Landing the fix regresses **`fused_gate_up_swiglu_rmsnorm_fp32_gamma_is_bit_exact`** (pristine `main`: 3 failed/27 passed → with fix: 4 failed/26 passed).

**Widening that DRY guard's allowlist to get a green test is NOT an acceptable resolution** — the guard exists to keep native and ORT decode on one shared policy, including against us. The invariant needs updating by **whoever owns it**, alongside their in-flight Estrin bit-exactness work (tracked in #1305). Once "capture-safe ⟺ M==1" is generalised to "capture-safe for the looped small-M path", this fix rebases and lands.

## Why this matters beyond batching (priority argument)

This fixed per-step penalty gates the **largest lever in the MoE offload line**. Another agent measured batching cutting expert-transfer **bytes/token by 61% (W=2→8)** — but that saving is in *bytes only* and **cannot convert to tokens/sec** while a flat ~11–22 ms per-step penalty stands: a route-aware scheduler would optimise the numerator while the denominator is broken. After this fix the step becomes **~1.4 ms fixed + ~1.9 ms/row** — a normal linear model in which expert-aware batching pays. (Full record: `docs/benchmarks/2026-08-19-batch-decode-mge2-capture-segmentation.md`, PRs #1324 / #1327.)

## Ask

- Owner of the capture-safe invariant (`matmul_nbits.rs:14561`/`:14962`) to update it alongside the Estrin fix (#1305), so the looped small-M SwiGLU path may be advertised capture-safe.
- Do **not** widen the guard as a shortcut.
- Once the invariant moves, `squad/batch-swiglu-capture-fix` rebases + lands for a measured 4.9× at batch-2.

Cross-refs: #1305 (owning capture-safe / Estrin failures), #1324 (diagnostic instrument, merged), #1327 (fixed-cost MoE input, merged), #1312 (the looped-GEMV pattern this mirrors).

**Comment by justinchuby** [comment]
## Landed the byte-identical half in #1404; plain half hard-stopped here

The blocker (the `only M=1 decode may be advertised capture-safe` invariant) was resolved by **measuring the ULP bound** instead of arguing it, then applying the decision rule.

### The outstanding measurement — ULP sweep of the fused decode GEMV vs the two-op reference

| path | M==1 (apples-to-apples decode) | M>1 |
|---|---|---|
| **rmsnorm-prologue** | **0 ULP** (0/60 cases) — byte-identical | ≤3 ULP (GEMV-vs-GEMM reduction order) |
| **plain** (no rmsnorm) | **2 ULP** (8/56 cases) | — |

The prior "1 ULP" claim was **refuted**: the plain fused decode GEMV is up to 2 ULP off the two-op reference **even at M==1**. The rmsnorm path, by contrast, is genuinely byte-identical at M==1.

### Decision rule → hard-stop the plain half, land the rmsnorm half

Because the plain path exceeds 1 ULP at M==1, per the rule I landed **only** the byte-identical rmsnorm half:

- **Landed (#1404):** M>1 rmsnorm gate/up-SwiGLU decode routes through a per-row capture-safe `M==1` GEMV loop (gated to `gamma.is_some()`). This is the resident-model decode path (the skip-rmsnorm fusion folds the pre-MLP RMS norm into the gate/up node for Qwen/Llama/Phi), so the measured **25 → 1 segment / 5.15 ms @ batch-2 (4.9×)** win is preserved. Byte-identical per-row == single-stream is guarded by a new test.
- **Hard-stopped (this issue):** the plain (no-rmsnorm) gate/up-SwiGLU decode. Its capture-safe invariant stays `m == 1`; plain M>1 stays on the prefill/Marlin path. Reason: 2 ULP from the two-op reference at M==1 → routing it through the M>1 loop would move plain-path logits >1 ULP. A `<=2` ULP sweep documents this in the code.

### Related

- Marlin already staled the `m == 1` invariant by default and its split-K reds three tests on pristine `main` — filed as #1405 (not fixed).
- Corrected #1305: the three matmul_nbits reds are Marlin split-K artifacts (toggle `ONNX_GENAI_MARLIN_M_GT_1=0`), not a rounding floor.

Closing as resolved by #1404 for the rmsnorm half. The plain-half hard-stop is recorded above; reopen/track separately if a deterministic ≤1 ULP plain decode path becomes available (e.g. a byte-identical Marlin or a plain decode GEMV proven ≤1 ULP by a future sweep).

**Comment by justinchuby** [comment]
**Correction (terminology):** my close comment above called the Marlin divergence "split-K". That is inaccurate — the default M>1 divergence is the **direct** Marlin int4 tensor-core GEMM (`launch_marlin_gemm`), not split-K. Split-K (`ONNX_GENAI_MARLIN_SPLITK`) is a separate, default-OFF, documented-deterministic path that does not engage under the default config. The toggle and conclusions are unchanged (`ONNX_GENAI_MARLIN_M_GT_1=0` flips the three tests red→green). #1405 has the precise write-up.

**Comment by justinchuby** [comment]
### e2e follow-up: fix is correct but gated OFF by default for small resident models

Completed the outstanding end-to-end measurement. On current `main` (RTX 4060, CUDA 13.1), the fix from #1404 is **proven correct** (segments 25→1, byte-identical when it engages), but it **does not engage by default** for the resident Qwen 0.5B family: their gate/up node runs the plain path (`gamma=None`) because the RMSNorm-gamma fold is gated off below hidden 1280 (qwen05b hidden=896). So the ~11–22 ms/step penalty this issue tracked is **still present by default** for those models.

Segments (default main): M=1→1, **M=2/4/8→25** (seam `MatMulNBits[KernelCaptureUnsupported]×24`). Forcing the fold on (`ONNX_GENAI_RMSNORM_MIN_HIDDEN=512`) collapses them to 1 at all M and yields the predicted ~1.4 ms fixed + ~1.9 ms/row model.

Root cause + proposed batch-aware fusion gate tracked in **#1421**. The fix landed here is not being reverted — it is the right fix; the follow-up is to let it engage for small models.
